### ISSUE TITLE
Toolbar Refactoring - Design Specs

### ISSUE THREAD

**Issue by mceledonia** [patternfly/patternfly-next#1528]
# Toolbar Refactoring

We decided to refactor the toolbar to a more simple and generic component in order to increase flexibility and improve its utility in a wider range of use cases.

### Toolbar Base
- The toolbar is 36px tall to match the height of our dropdowns and buttons. 
- All items not this height (icons, text) should be centered vertically within the height.
- There should no spacers on the left and right edges of the toolbar if there are already padding spacers being used with whatever component the toolbar is placed within (e.g. cards).

<img width="380" alt="image" src="https://user-images.githubusercontent.com/30935938/53754773-f181d880-3e82-11e9-9500-214b9f0cca22.png">

### Toolbar Items: Group
Items can be grouped within the toolbar. Items within a group have 16px spacing between them. Items that are made up of multiple parts (for example the filter dropdown and input group) are treated as one item in regards to the spacing. 

Below are examples of: a group of icons, a group of buttons, a group of icons and buttons. Note that no matter what the item is, it always has 16px of spacing on both sides.

<img width="336" alt="image" src="https://user-images.githubusercontent.com/30935938/53754951-5d644100-3e83-11e9-9d98-8bc5518a0057.png">

### Toolbar Items: Multiple Groups
A toolbar can have more than one group of items. Between groups there should be 24px of spacing.
Below is a group of icons next to a group of buttons. Note the cyan 24px spacer between them.

<img width="394" alt="image" src="https://user-images.githubusercontent.com/30935938/53755209-f3986700-3e83-11e9-960d-e751236c06a1.png">

### Toolbar Items: Single, ungrouped icons are treated as a group.
A single item that doesn't belong to a larger group is treated as a group of its own. It uses a 24px spacer on either side.
Below is a single button next to a group of icons.

<img width="373" alt="image" src="https://user-images.githubusercontent.com/30935938/53755331-4114d400-3e84-11e9-9b6b-f529fc23f2c2.png">

### Example: Card with toolbar
Below is an example of a toolbar within a card. This card has a header and a body section, the toolbar should always be placed in the header. The toolbar has a single button alongside a group containing action icons. Note how the toolbar has no padding on the right side as the padding is taken care of by the card component padding.

<img width="448" alt="image" src="https://user-images.githubusercontent.com/30935938/53755453-97821280-3e84-11e9-8419-7a6f309b98dd.png">

<img width="446" alt="image" src="https://user-images.githubusercontent.com/30935938/53755587-de700800-3e84-11e9-8bbe-7ebf78358e2f.png">

### FYI
@mcoker @rachael-phillips @kybaker @mcarrano

**Comment by mcarrano** [comment]
@mceledonia agree that this refactoring makes sense to create a more flexible toolbar as we previously discussed. 

@rachael-phillips @mcoker do we have this on the roadmap as something to deliver for April release? Sounds to me like this would be a breaking change so it would be good to get in sooner rather than later. Does this refactoring also require rework for React?

**Comment by mcoker** [comment]
@mcarrano We could introduce it as a new component, then deprecate the old one, to add the component as a step, then at a later step, remove the old one. This will also involve breaking changes in at least the page layout (masthead buttons/menus, though not sure how the new masthead design might impact this), and potentially the form layout for the action row of buttons, whenever we decide to implement those changes.

**Comment by mcarrano** [comment]
@mcoker and I discussed this and wanted to offer the following as a potential path forward. 

We can keep the existing Toolbar component which was designed for a very specific use case to sit on top of a table or list view. This is useful for anyone creating a standardized data view. We can then take this "refactored toolbar" into a new component that can be used in a more generalized way - for layout out masthead utility items, actions, or filters within a card. It can also be useful if someone needs to roll their own toolbar that does not fit neatly into one of the templates defined in the current Toolbar component.

I think the only challenge is what to call the new thing. It's still technically a toolbar, but we want to differentiate it. We were thinking maybe "Action Bar," but are open to ideas.

Thoughts @mceledonia @rachael-phillips ?

**Comment by rachael-phillips** [comment]
My question would the new Action bar cause breaking changes?

**Comment by mcoker** [comment]
@rachael-phillips adding the new component wouldn't cause a breaking change. And using it in the page header (masthead utility items) wouldn't necessarily be a breaking change if we add support for the toolbar to be used and don't remove the old styles that support the current layout. Then we can go back later in a breaking changes release and remove support for the current layout and force users to use the toolbar there instead. I imagine we would also eventually replace the current toolbar with this toolbar (or action bar) as well.

**Comment by rachael-phillips** [comment]
What do you think @mceledonia ?

**Comment by mceledonia** [comment]
Additional proposal for a vertical divider between item groups, mainly utilized when there are two or more groups of the same item types next to each other (e.g. two groups of icons next to each other as seen below). In other words this is to be used when just the spacing is not enough of a visual separator between groups.

![image](https://user-images.githubusercontent.com/30935938/54606782-1f3e5400-4a23-11e9-8288-c0cd88c71e13.png)

@mcoker @kybaker

**Comment by katierik** [comment]
Re: the order of the buttons up on this toolbar. Everywhere we're doing "Primary, Secondary, Tertiary" (on the bottom of forms, in wizards, etc) and in this toolbar design its "Tertiary, Secondary, Primary". 

It would be better to have the buttons in the same order all the time to reduce confusion & enhance consistency. 
(Unrelated, I have *opinions* about the ordering of the buttons in general, but consistency is key!)

**Comment by mcarrano** [comment]
I think @katierik makes a good point. On toolbar actions we have always had the primary action as the first (leftmost) button. I think this was an unintentional change. @mceledonia can you update your mock to avoid any confusion?

**Comment by mceledonia** [comment]
I agree 100% wherever we land, we should be consistent. It sounds like we might need to revisit button placement and have a larger discussion around it. Maybe we can find some time this week to meet and discuss?

**Comment by mattnolting** [comment]
Hey all, thought I'd add my thoughts here. 

I think that we're interpreting what is a component that uses `display: flex;` as a unique component with its own unique layout. It is correct that this design shows a toolbar, the approach to layout should be consistent with anything else that is a flexbox layout. Therefore, I think it's a good idea to create a `toolbar` component that uses the `pf-l-flex` layout.

Adding markup/grouping items to control margin seems counterintuitive and unnecessary to me and I've heard the same feedback from others (on my team). Like, there has to be more than 1 thing to make a group. Rather, a modifier could be added to an item or conditional css for that item could be added..

**Comment by mcarrano** [comment]
@mattnolting should this issue be closed?

**Comment by mattnolting** [comment]
@mcarrano No, it should be left open. The `pf-l-flex` layout has been updated and I'm in the process of applying it to toolbar. Summit derailed development, but picking it back up now.

**Comment by rachael-phillips** [comment]
@mattnolting what's the status on this issue and how is it going? :)

**Comment by mattnolting** [comment]
@rachael-phillips After meeting w/team members about approach last week, we came to the conclusion that instead of applying selectors `.pf-l-flex` and modifiers to the toolbar (and any other component that relies on our standardized flex layout) effectively creates a blueprint for, rather than a complete, autonomous component and would ship with a much larger module size. As a result, I'm in the process of declaration abstraction by converting modifiers to rely on sass mixins/placeholders. This will allow us to ship only the required amount of css to support a given component, rather than a component and a complete layout. 

The bulk of the css work is done and am working out some details. Documentation has been started with examples. It's going well :).

**Comment by rachael-phillips** [comment]
Thanks so much @mattnolting for the update!

**Comment by mcarrano** [comment]
Adding link to Toolbar Component System design doc for future reference: https://docs.google.com/document/d/12oxvXul8BxxjotRWSpJefyuiqcSm_gOYscEOXZXaqaI/edit?usp=sharing

**Comment by mattnolting** [comment]
Toolbar examples (should be able to support these)
https://marvelapp.com/project/4205050/
https://marvelapp.com/project/4204560/
https://marvelapp.com/project/4148169/
https://marvelapp.com/project/4100287/
